pyglmnet/pyglmnet.py

```diff
@@ -38,6 +38,175 @@ def set_log_level(verbose):
+        L1penalty = 0.0
+        for group_id in group_ids:
+            if group_id != 0:
+                L1penalty += \
+                    np.linalg.norm(beta[group == group_id], 2)
+        L1penalty += np.linalg.norm(beta[group == 0], 1)
+    return L1penalty
+
+
+def _loss(distr, alpha, Tau, reg_lambda, X, y, eta, group, beta):
+    """Define the objective function for elastic net."""
+    L = _logL(distr, beta[0], beta[1:], X, y, eta)
+    P = _penalty(alpha, beta[1:], Tau, group)
+    J = -L + reg_lambda * P
+    return J
+
+
+def _grad_L2loss(distr, alpha, reg_lambda, X, y, Tau, eta, beta):
+    """The gradient."""
+
+    n_samples = np.float(X.shape[0])
+
+    if Tau is None:
+        Tau = np.eye(beta[1:].shape[0])
+    InvCov = np.dot(Tau.T, Tau)
+
+    z = beta[0] + np.dot(X, beta[1:])
+    s = expit(z)
+
+    if distr == 'softplus':
+        q = _qu(distr, z, eta)
+        grad_beta0 = 1. / n_samples * (np.sum(s) - np.sum(y * s / q))
+        grad_beta = 1. / n_samples * \
+            (np.transpose(np.dot(np.transpose(s), X) -
+                          np.dot(np.transpose(y * s / q), X))) + \
+            reg_lambda * (1 - alpha) * \
+            np.dot(InvCov, beta[1:])
+
+    elif distr == 'poisson':
+        q = _qu(distr, z, eta)
+        grad_beta0 = np.sum(q[z <= eta] - y[z <= eta]) + \
+            np.sum(1 - y[z > eta] / q[z > eta]) * \
+            np.exp(eta)
+        grad_beta0 *= 1. / n_samples
+
+        grad_beta = np.zeros([X.shape[1], 1])
+        selector = np.where(z.ravel() <= eta)[0]
+        grad_beta += np.transpose(np.dot((q[selector] - y[selector]).T,
+                                         X[selector, :]))
+        selector = np.where(z.ravel() > eta)[0]
+        grad_beta += np.exp(eta) * \
+            np.transpose(np.dot((1 - y[selector] / q[selector]).T,
+                                X[selector, :]))
+        grad_beta *= 1. / n_samples
+        grad_beta += reg_lambda * (1 - alpha) * \
+            np.dot(InvCov, beta[1:])
+
+    elif distr == 'gaussian':
+        grad_beta0 = 1. / n_samples * np.sum(z - y)
+        grad_beta = 1. / n_samples * \
+            np.transpose(np.dot(np.transpose(z - y), X)) \
+            + reg_lambda * (1 - alpha) * \
+            np.dot(InvCov, beta[1:])
+
+    elif distr == 'binomial':
+        grad_beta0 = 1. / n_samples * np.sum(s - y)
+        grad_beta = 1. / n_samples * \
+            np.transpose(np.dot(np.transpose(s - y), X)) \
+            + reg_lambda * (1 - alpha) * \
+            np.dot(InvCov, beta[1:])
+
+    elif distr == 'multinomial':
+        # this assumes that y is already as a one-hot encoding
+        pred = _qu(distr, z, eta)
+        grad_beta0 = -1. / n_samples * np.sum(y - pred, axis=0)
+        grad_beta = -1. / n_samples * \
+            np.transpose(np.dot(np.transpose(y - pred), X)) \
+            + reg_lambda * (1 - alpha) * \
+            np.dot(InvCov, beta[1:])
+
+    n_classes = y.shape[1] if distr == 'multinomial' else 1
```

Comment:
@pavanramkumar is the grad for `multinomial` case also going to be 2D? I think you can vectorize it and then reshape it outside of the loss function